taskmanager/server.go

```diff
@@ -0,0 +1,155 @@
+package taskmanager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/tcsc/levity/api"
+	"github.com/tcsc/levity/registry"
+	"github.com/tcsc/levity/task"
+)
+
+// Error is an error code describing things that can go wrong with the
+// TaskManager API service.
+type Error int
+
+const (
+	// NoSuchTask indicates that the operation was targeted at a task
+	// that does not exist.
+	NoSuchTask Error = 1
+)
+
+func (e Error) Error() string {
+	switch e {
+	case NoSuchTask:
+		return "No such task"
+	}
+	return fmt.Sprintf("Unknown error code: %d", int(e))
+}
+
+// TaskRegistry abstracts out the desinition of the task0id-to-task mapping, so
+// we can mock it ut for testing
+type TaskRegistry interface {
+	Register(ctx context.Context, t *task.Task) (string, error)
+	Lookup(id string) *task.Task
+}
+
+// Server is an implementation of the TaskManager API
+type Server struct {
+	api.UnimplementedTaskManagerServer
+	registry TaskRegistry
+}
+
+// New creates and initialises a new Server with default settings
+func New() *Server {
+	return withRegistry(registry.New())
+}
+
+// withRegistry creates and initialises a new Server with a custom task registry
+// implementation
+func withRegistry(registry TaskRegistry) *Server {
+	return &Server{registry: registry}
+}
+
+// StartTask attempts to start and register a task with the task manager.
+func (server *Server) StartTask(ctx context.Context, req *api.StartTaskRequest) (*api.StartTaskResponse, error) {
+	t := task.New(
+		req.GetBinary(),
+		req.GetWorkingDir(),
+		req.GetEnvironment(),
+		req.GetArgs()...)
+
+	// Start the task *before* registering it, as we don't want to expose the task
+	// to RPC calls from the outside world until the task is in a state that can
+	// handle it.
+	err := t.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	// Register the task with the registry, which will make it available to other
+	// RPC calls
+	id, err := server.registry.Register(ctx, t)
+	if err != nil {
+		killTask(t) // Don't leak the task if we fail to register it
+		return nil, err
+	}
+
+	// Give the caller a handle to their task
+	return &api.StartTaskResponse{
+		TaskId: &api.TaskHandle{Id: id},
+	}, nil
+}
+
+// FetchLogs extracts and returns the collected stdout & stderr data from the
+// task
+func (server *Server) FetchLogs(
+	ctx context.Context, req *api.FetchLogsRequest) (*api.FetchLogsResponse, error) {
+
+	task := server.registry.Lookup(req.TaskId.Id)
```

Comment:
More of a nit, but looking at this I think the contract would be a little bit cleaner if the registry's Lookup method just returned a not found error itself rather than translating "nil" to "not found" in server.

Reply:
I'm on the fence with this one. I can see what you're saying, but I also like the separation of concerns, where the registry is an internal component that does what it does, while the `Server` is responsible for what errors get propagated to the client.

 It feels a bit like leaking implementation details.

I'm going to leave it as-is for now, but happy for further guidance.

Reply:
This is a minor issue, no need to make changes on this, but since you asked, take a look at the following article.

https://en.wikipedia.org/wiki/Semipredicate_problem

I know this is one of the things I disliked about C myself.